pkg/volume/csi/csi_plugin.go

```diff
@@ -785,13 +798,61 @@ func (p *csiPlugin) skipAttach(driver string) (bool, error) {
 	return false, nil
 }
 
-// getDriverMode returns the driver mode for the specified spec: {persistent|ephemeral}.
+// supportsVolumeMode looks up CSIDriver object associated with the
+// driver name to determine if the driver supports the given volume
+// mode. An error indicates that it isn't supported and explains why.
+func (p *csiPlugin) supportsVolumeMode(driver string, volumeMode csiVolumeMode) error {
+	driverMode, isDefault, err := p.lookupDriverMode(driver)
+	switch {
+	case err != nil:
+		return err
+	case volumeMode == persistentVolumeMode && (driverMode == storage.PersistentDriverMode || driverMode == storage.CombinedDriverMode) ||
+		volumeMode == ephemeralVolumeMode && (driverMode == storage.EphemeralDriverMode || driverMode == storage.CombinedDriverMode):
+		return nil
+	case isDefault:
+		// Using the same error message might be confusing because 'which only supports "persistent"' would
+		// be wrong for a driver that merely doesn't specify that it supports ephemeral.
+		return fmt.Errorf("volume mode %q not supported by driver %s, no CSIDriver object or no Mode specified", volumeMode, driver)
+	default:
+		return fmt.Errorf("volume mode %q not supported by driver %s which only supports %q", volumeMode, driver, driverMode)
+	}
+}
+
+// lookupDriverMode looks up CSIDriver object associated with the driver name and
+// returns what it finds there or the "persistent" default.
+func (p *csiPlugin) lookupDriverMode(driver string) (storage.CSIDriverMode, bool, error) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIDriverRegistry) {
+		// No further information, assume the default applies.
+		return storage.PersistentDriverMode, true, nil
+	}
+
+	kletHost, ok := p.host.(volume.KubeletVolumeHost)
+	if ok {
+		kletHost.WaitForCacheSync()
+	}
+
+	if p.csiDriverLister == nil {
+		return "", false, errors.New("CSIDriver lister does not exist")
+	}
+	csiDriver, err := p.csiDriverLister.Get(driver)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			// No driver info, assume the default.
+			return storage.PersistentDriverMode, true, nil
+		}
+		return "", false, err
+	}
+	if csiDriver.Spec.Mode == nil {
+		return storage.PersistentDriverMode, true, nil
```

Comment:
Should `false` be returned since driver mode is not set.  the boolean needs to be clarified I suppose.